src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -156,6 +201,492 @@ MultiRouterPlanCreate(Query *originalQuery, Query *query,
+		{
+			colocationId = currentColocationId;
+		}
+
+		if (colocationId != currentColocationId)
+		{
+			tablesAreColocated = false;
+			break;
+		}
+	}
+
+
+	if (!tablesAreColocated)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("all participating tables should be colocated")));
+	}
+}
+
+
+/*
+ * ErrorIfInsertPartitionColumnDoesNotMatchSelect checks whether the INSERTed table's
+ * partition column value matches with the any of the SELECTed table's partition column.
+ */
+static void
+ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query)
+{
+	ListCell *targetEntryCell = NULL;
+	uint32 rangeTableId = 1;
+	RangeTblEntry *insertRte = linitial(query->rtable);
+	RangeTblEntry *subqueryRte = lsecond(query->rtable);
+	Query *subquery = subqueryRte->subquery;
+	Oid insertRelationId = insertRte->relid;
+	Var *insertPartitionColumn = PartitionColumn(insertRelationId, rangeTableId);
+	bool partitionColumnsMatch = false;
+
+	foreach(targetEntryCell, query->targetList)
+	{
+		TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+
+		if (IsA(targetEntry->expr, Var))
+		{
+			Var *insertVar = (Var *) targetEntry->expr;
+			AttrNumber originalAttrNo = get_attnum(insertRelationId,
+												   targetEntry->resname);
+			TargetEntry *subqeryTargetEntry = NULL;
+
+			if (originalAttrNo != insertPartitionColumn->varattno)
+			{
+				continue;
+			}
+
+			subqeryTargetEntry = list_nth(subquery->targetList,
+										  insertVar->varattno - 1);
+
+			if (!IsA(subqeryTargetEntry->expr, Var))
+			{
+				partitionColumnsMatch = false;
+				break;
+			}
+
+			if (!IsPartitionColumnRecursive(subqeryTargetEntry->expr, subquery))
+			{
+				partitionColumnsMatch = false;
+				break;
+			}
+
+			partitionColumnsMatch = true;
+			break;
+		}
+	}
+
+	if (!partitionColumnsMatch)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("SELECT query should return bare partition column on "
+							   "the same ordinal position with INSERT query's partition "
+							   "column")));
+	}
+}
+
+
+/*
+ * AddHiddenPartitionColumnParameter() can only be used with
+ * INSERT ... SELECT queries.
+ *
+ * If the input query is not INSERT .. SELECT the function errors-out.
+ */
+void
+AddHiddenPartitionColumnParameter(Query *originalQuery)
+{
+	Query *subquery = NULL;
+	Query *referencedSubquery = NULL;
+	RangeTblEntry *subqueryEntry = NULL;
+
+	if (!InsertSelectQuery(originalQuery))
+	{
+		elog(ERROR, "Only INSERT .. SELECT queries can be modified");
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot perform distributed planning for the given"
+							   " modification")));
+	}
+
+	/* TODO: once CTEs are present, this does not work */
+	subqueryEntry = (RangeTblEntry *) list_nth(originalQuery->rtable, 1);
+	subquery = subqueryEntry->subquery;
+
+	referencedSubquery = FirstQueryReferencingDistributedTable(subquery);
+
+	AddHiddenParameterToFirstTable(referencedSubquery);
+}
```

Comment:
Maybe I'm missing something, but the planner will push down the qual in cases it's safe and it knows how to.  But it's not generally if there's grouping involved, that's why it probably isn't working in this case. But in cases without barriers for pushing down it should work.